gtsfm/densify/mvsnets/methods/PatchmatchNet/models/patchmatch.py

```diff
@@ -0,0 +1,891 @@
+
+            # evaluation, outputs regressed depth map
+            depth_sample, score = self.evaluation(
+                ref_feature,
+                src_features,
+                ref_proj,
+                src_projs,
+                depth_sample,
+                depth_min,
+                depth_max,
+                iter,
+                eval_grid,
+                weight,
+                view_weights,
+            )
+
+            depth_sample = depth_sample.unsqueeze(1)
+            depth_samples.append(depth_sample)
+
+        return depth_samples, score, view_weights
+
+
+class SimilarityNet(nn.Module):
+    """Similarity Net
+    first, do convolution on aggregated cost among all the source views
+    second, perform adaptive spatial cost aggregation to get final cost
+    """
+
+    def __init__(self, G: int, neighbors: int = 9):
+        """initialize
+        Args:
+            G: group number
+            neighbors: number of neighbors
+        """
+        super(SimilarityNet, self).__init__()
+        self.neighbors = neighbors
+
+        self.conv0 = ConvBnReLU3D(G, 16, 1, 1, 0)
+        self.conv1 = ConvBnReLU3D(16, 8, 1, 1, 0)
+        self.similarity = nn.Conv3d(8, 1, kernel_size=1, stride=1, padding=0)
+
+    def forward(self, x1: torch.Tensor, grid: torch.Tensor, weight: torch.Tensor) -> torch.Tensor:
+        """forward function
+        Args:
+            x1: [B, G, Ndepth, H, W], aggregated cost among all the source views with pixel-wise view weight
+            grid: position of sampling points in adaptive spatial cost aggregation
+            weight: weight of sampling points in adaptive spatial cost aggregation, combination of
+                feature weight and depth weight
+        """
+
+        batch, G, num_depth, height, width = x1.size()
+
+        x1 = self.similarity(self.conv1(self.conv0(x1))).squeeze(1)
+
+        x1 = F.grid_sample(x1, grid, mode="bilinear", padding_mode="border")
+
+        # [B,Ndepth,9,H,W]
+        x1 = x1.view(batch, num_depth, self.neighbors, height, width)
+
+        return torch.sum(x1 * weight, dim=2)
+
+
+class FeatureWeightNet(nn.Module):
+    """FeatureWeight Net
+    adaptive spatial cost aggregation
+    weight based on similarity of features of sampling points and center pixel
+    """
+
+    def __init__(self, num_feature: int, neighbors: int = 9, G: int = 8):
+        """initialize
+        Args:
+            num_features: number of features
+            neighbors: number of neighbors
+            G: number of groups
+        """
+        super(FeatureWeightNet, self).__init__()
+        self.neighbors = neighbors
+        self.G = G
+
+        self.conv0 = ConvBnReLU3D(G, 16, 1, 1, 0)
+        self.conv1 = ConvBnReLU3D(16, 8, 1, 1, 0)
+        self.similarity = nn.Conv3d(8, 1, kernel_size=1, stride=1, padding=0)
+
+        self.output = nn.Sigmoid()
+
+    def forward(self, ref_feature: torch.Tensor, grid: torch.Tensor) -> torch.Tensor:
+        """forward

```

Comment:
This docstring could be clearer -- feed what type of input through what type of network? What is the meaning of the output?